### PR TITLE
[development] add configManager to kafkaWS

### PR DESCRIPTION
**Pull request by jonaphael**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR?** (Link to an issue, e.g. #99999)

* **Other information**:

### CODE REVIEW

File: subscription-engine/kafka-ws/config/default.conf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# SERVER
+server.host=0.0.0.0
+server.port:integer=8080
+server.tls:boolean=false
+server.ca=/opt/kafka-ws/certs/ca-cert.pem
+server.key=/opt/kafka-ws/certs/server-key.pem
+server.cert=/opt/kafka-ws/certs/server-cert.pem
+server.jwt_exp_time:boolean=false
+server.connection_max_life_time:integer=7200
+server.request_cert:boolean:false
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/config/default.conf:32]: Underscores should not be used as default, only in some special cases. For an example of special case, check the V2K config.

File: subscription-engine/kafka-ws/config/default.conf [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# LOGGER
+logger.transports.console.level=info
+logger.verbose:boolean=false
+
+logger.file.enable:boolean=false
+logger.file.level=info
+logger.file.filename=kafka-ws-logs-%DATE%.log
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/config/default.conf:7]: I would change the names to:
```
logger.transports.file.*
```
This way we keep the same pattern as the console configuration.

File: subscription-engine/kafka-ws/index.js [hunk 3]
```diff
@@ -1,36 +1,46 @@
...
   const options = {
-    cert: fs.readFileSync(serverCfg.tls_cert_file),
-    key: fs.readFileSync(serverCfg.tls_key_file),
-    ca: [fs.readFileSync(serverCfg.tls_ca_file)],
+    cert: fs.readFileSync(config.server.cert),
+    key: fs.readFileSync(config.server.key),
+    ca: [fs.readFileSync(config.server.ca)],
     rejectUnauthorized: true,
-    requestCert: serverCfg.request_cert,
+    requestCert: config.server.request_cert,
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/index.js:43]: Don't forget the names, this one, as I said in another comment in the default.conf, is not conforming to the patterns of the ConfigManager module.

File: subscription-engine/kafka-ws/index.js [hunk 4]
```diff
@@ -1,36 +1,46 @@
...
+if (config.server.tls) {
   logger.info('Initializing the HTTP server (Using TLS Protocol)...');
   const options = {
-    cert: fs.readFileSync(serverCfg.tls_cert_file),
-    key: fs.readFileSync(serverCfg.tls_key_file),
-    ca: [fs.readFileSync(serverCfg.tls_ca_file)],
+    cert: fs.readFileSync(config.server.cert),
+    key: fs.readFileSync(config.server.key),
+    ca: [fs.readFileSync(config.server.ca)],
     rejectUnauthorized: true,
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/index.js:42]: I would add this into the default.conf too.

File: subscription-engine/kafka-ws/app/Redis/RedisExpireMgmt.js [hunk 5]
```diff
@@ -22,12 +26,7 @@ class RedisExpireMgmt {
     //   ...
     // };
 
-    const redisOptions = {
-      host: redisConfig.host,
-      port: redisConfig.port,
-      db: redisConfig.database,
-      // tls: tls_options
-    };
+    const redisOptions = { ...this.config };
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/app/Redis/RedisExpireMgmt.js:29]: This is not necessary. You already have the object in the `this.config` variable.

File: subscription-engine/kafka-ws/app/Redis/RedisManager.js [hunk 6]
```diff
@@ -16,12 +16,10 @@ class RedisManager {
...
 
+    this.config = ConfigManager.getConfig(KAFKA_WS_CONFIG_LABEL).redis;
+
     // TODO: Implement "retry_strategy"
-    const redisOptions = {
-      host: redisConfig.host,
-      port: redisConfig.port,
-      db: redisConfig.database,
-    };
+    const redisOptions = { ...this.config };
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/app/Redis/RedisManager.js:22]: This is not necessary. You already have the object in the `this.config` variable.

File: subscription-engine/kafka-ws/app/WebsocketTarball.js [hunk 7]
```diff
@@ -14,7 +14,11 @@ const {
   isObjectEmpty,
   addTimeFromNow,
 } = require('./Utils');
-const { server: serverConfig } = require('./Config');
+
+// Loading the configurations with the configManager
+const KAFKA_WS_CONFIG_LABEL = 'KAFKA_WS';
+const config = ConfigManager.getConfig(KAFKA_WS_CONFIG_LABEL);
+const serverConfig = { ...config.server };
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/app/WebsocketTarball.js:21]: There is no need to apply the object destructuring here.

**Review by eduardogmisiuk — COMMENTED** [review]
Nice! There are some adjustments to be made, but overall it is very nice.

File: subscription-engine/kafka-ws/jest.setup.js [hunk 8]
```diff
@@ -1,8 +1,10 @@
...
 
-Logger.setTransport('console', {
-  level: appCfg.log.log_console_level,
-});
-Logger.setVerbose(appCfg.log.log_verbose);
+ConfigManager.loadSettings(KAFKA_WS_CONFIG_LABEL, 'default.conf');
+
+const config = ConfigManager.getConfig(KAFKA_WS_CONFIG_LABEL);
+
+Logger.setTransport('console', { level: config.logger['logger.transports.console.level'] });
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/jest.setup.js:9]: Isn't it `transports.console.level` instead of `logger.transports.console.level`?

**Review by eduardogmisiuk — COMMENTED** [review]
Nice, I left only one comment. And also, have you checked both examples? I saw you changed one, was the other one OK?

File: subscription-engine/kafka-ws/app/Kafka/KafkaConsumer.js [hunk 9]
```diff
@@ -13,8 +14,10 @@ class KafkaConsumer {
   * Instance KafkaConsumer
   */
   constructor() {
+    this.consumerConfig = ConfigManager.getConfig(KAFKA_WS_CONFIG_LABEL).consumer;
+
     logger.debug('constructor: Instance KafkaConsumer');
-    this.consumer = new Consumer(consumerConfig);
+    this.consumer = new Consumer(this.consumerConfig);
```
Comment (rascaraficci) [inline subscription-engine/kafka-ws/app/Kafka/KafkaConsumer.js:20]: I think we have a problem here. Look at https://github.com/dojot/dojot-microservice-sdk-js/blob/68950d6d71b0a3b16ae5cb5f560fdbab2dbbf7a9/examples/consumer/sample.js#L13 . The first parameter is consumer settings, but the second one is topic settings. We are passing `consumer.auto.offset.reset=largest` as a consumer property, but it is a topic property.

Other services have this same problem.

**Review by rascaraficci — CHANGES_REQUESTED** [review]
Great job. I left few comments.

**Comment by mprevide** [comment]
Look the unit tests

`Test Suites: 1 failed, 8 passed, 9 total
Tests:       1 failed, 112 passed, 113 total
Snapshots:   0 total
Time:        12.716s
Ran all test suites.
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
Error: The operation was canceled.`

**Comment by jonaphael** [comment]
> The unit tests
> 
> ```
> Test Suites: 1 failed, 8 passed, 9 total
> Tests:       1 failed, 112 passed, 113 total
> Snapshots:   0 total
> Time:        12.716s
> Ran all test suites.
> Jest did not exit one second after the test run has completed.
> 
> This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
> Error: The operation was canceled.
> ```

Thanks @mprevide the error was for a missing parameter on a mock configuration.

File: subscription-engine/kafka-ws/jest.setup.js (hunk 8 above)
Comment (jonaphael) [reply subscription-engine/kafka-ws/jest.setup.js:9]: Fixed!

**Review by rascaraficci — APPROVED** [review]
It looks good to me!

File: subscription-engine/kafka-ws/README.md [hunk 10]
```diff
@@ -292,30 +292,74 @@ operators](###applying-conditions-where) for more info
...
-REDIS_PORT             | Redis port                                                                                  | 6379                                | number                        |
-REDIS_DATABASE         | Redis database                                                                              | 1                                   | number                        |
-KAFKA_WS_REQUEST_CERT  | Whether the service should request for a client certificate or not                          | false                               | boolean                       |
-
-__NOTE THAT__ a websocket connection is closed by the server when certain conditions are met. If KAFKA_WS_JWT_EXP_TIME is set to true, the server will consider this value for closing the connection if it is greater than the KAFKA_WS_MAX_LIFE_TIME. If KAFKA_WS_JWT_EXP_TIME is set to false and KAFKA_WS_MAX_LIFE_TIME is set to -1, the server will never close a connection by its duration.
+You can select the configuration file via the `KAFKA_WS_APP_USER_CONFIG_FILE` variable. Its default value is `production.env`
```
Comment (eduardogmisiuk) [inline subscription-engine/kafka-ws/README.md:295]: I see that you got this part from the V2K README. I updated it the other day adding this:
```
Check the [config directory](./config) for the user configurations that are
available by default.
```

I think you could add it too.

**Review by eduardogmisiuk — COMMENTED** [review]
We're almost there! I left one improvement to the README.

File: subscription-engine/kafka-ws/README.md (hunk 10 above)
Comment (jonaphael) [reply subscription-engine/kafka-ws/README.md:295]: Great!

**Review by eduardogmisiuk — APPROVED** [review]
LGTM!

**Pull request merged**